Harvard Caselaw Access Project.js

```diff
@@ -0,0 +1,175 @@
+	"label": "Harvard Caselaw Access Project",
+	"creator": "Franklin Pezzuti Dyer",
+	"target": "^https://(.*\\.)*case\\.law",
+	"minVersion": "5.0",
+	"maxVersion": "",
+	"priority": 100,
+	"inRepository": true,
+	"translatorType": 4,
+	"browserSupport": "gcsibv",
+	"lastUpdated": "2024-01-15 23:41:24"
+}
+
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2024 Franklin Pezzuti Dyer
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+function detectWeb(doc, _url) {
+	if (doc.querySelectorAll('.case-container').length > 0) {
+		return 'case';
+	}
+	else if (getSearchResults(doc, true)) {
+		return 'multiple';
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('div.result-title > div > a');
+	for (let row of rows) {
+		let href = row.href;
+		Z.debug(href);
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+async function doWeb(doc, url) {
+	if (detectWeb(doc, url) == 'multiple') {
+		let items = await Zotero.selectItems(getSearchResults(doc, false));
+		if (!items) return;
+		for (let url of Object.keys(items)) {
+			await scrape(await requestDocument(url));
+		}
+	}
+	else {
+		await scrape(doc, url);
+	}
+}
+
+async function scrape(doc, url = doc.location.href) {
+	let apiUrl = attr(doc, "a[href*='api.case.law/v1/cases/']", 'href');
+	let caseText = await requestText(apiUrl);
+	let caseJson = JSON.parse(caseText);
+
+	let caseItem = new Zotero.Item("case");
+	caseItem.language = "en-US";
+	caseItem.url = url;
+
+	caseItem.caseName = caseJson.name;
+	caseItem.shortTitle = caseJson.name_abbreviation;
+	caseItem.court = caseJson.court.name;
+	caseItem.dateDecided = caseJson.decision_date;
+	caseItem.docketNumber = caseJson.docket_number;
+	caseItem.reporter = caseJson.reporter.full_name;
+	caseItem.reporterVolume = caseJson.volume.volume_number;
+
+	caseItem.complete();
+}
+
+/** BEGIN TEST CASES **/
+var testCases = [
+	{
+		"type": "web",
+		"url": "https://cite.case.law/am-samoa/2/3/",
+		"items": [
+			{
+				"itemType": "case",
+				"caseName": "PASA of FAGATOGO, Plaintiff v. FAIISIOTA of FAGANEANEA, Defendant",
```

Comment:
we'll probably have to leave this as is, but I'm not loving the all caps here, if we can get rid of them? Problem is that I don't see a way in which we wouldn't also get rid of (properly) all capsed acronyms that way.

Reply:
One brute-force solution occurs to me:
1. Split the title into words and detect which ones are all-caps.
2. For each of those words, search the case text body for the first (case-insensitive) match.
3. If an all-caps match is found, chances are that it's an acronym and should be left capitalized.
4. Otherwise, capitalize the word like a name. (I'm betting on the likelihood that if it's an acronym, it will appear again within the case text.)
This should deal with acronyms properly, but it involves searching an arbitrary amount of the case's text. Matches can probably be found early when they're present at all, so it may suffice to just look at the first couple paragraphs. Overkill?

Reply:
That's clever -- and this is on triggered actions (i.e. within `doWeb`) so we don't need to go crazy chasing efficiency (and it doesn't make additional requests to the site. Let's try it if you don't mind?